### PR TITLE
[AutoDiff upstream] Add SIL differentiability witnesses.

### PR DESCRIPTION
**Pull request by dan-zheng**
SIL differentiability witnesses are a new top-level SIL construct mapping
"original" SIL functions to derivative SIL functions.

SIL differentiability witnesses have the following components:
- "Original" `SILFunction`.
- SIL linkage.
- Differentiability parameter indices (`IndexSubset`).
- Differentiability result indices (`IndexSubset`).
- Derivative `GenericSignature` representing differentiability generic requirements (optional).
- JVP derivative `SILFunction` (optional).
- VJP derivative `SILFunction` (optional).
- "Is serialized?" bit.

```
sil_differentiability_witness hidden [parameters 0] [results 0] <T where T : Differentiable> @id : $@convention(thin) (T) -> T {
  jvp: @id_jvp : $@convention(thin) (T) -> (T, @owned @callee_guaranteed (T.TangentVector) -> T.TangentVector)
  vjp: @id_vjp : $@convention(thin) (T) -> (T, @owned @callee_guaranteed (T.TangentVector) -> T.TangentVector)
}
```

This patch adds the `SILDifferentiabilityWitness` data structure, with
documentation, parsing, and printing.

Resolves TF-911.

---

Todos:
- TF-1136: upstream `SILDifferentiabilityWitness` serialization.
- TF-1137: upstream `SILDifferentiabilityWitness` verification.
- TF-1138: upstream `SILDifferentiabilityWitness` SILGen from
  `@differentiable` and `@derivative` attributes.
- TF-20: robust mangling for `SILDifferentiabilityWitness` names.

### CODE REVIEW

**Review by rxwei — APPROVED** [review]

File: lib/SIL/SILPrinter.cpp
Comment (dan-zheng) [inline lib/SIL/SILPrinter.cpp]: Note: printing SIL differentiability witnesses before SIL functions may be slightly controversial.

This decision significantly simplifies SIL parsing: it allows us to avoid adding support for diff. witness forward declarations when parsing `differentiability_witness_function` instructions in SIL functions.

Other longer table-like SIL constructs are printed after SIL functions so that functions appear at the top. Diff. witnesses aren't as long as other table-like SIL constructs (up to four lines), so they don't take up too much space.

**Comment by dan-zheng** [comment]
@jckarter: would you like to review this PR, or suggest a reviewer? I'm asking you since you seem like a SIL code owner.

This PR is part of our effort to [upstream differentiable programming](https://bugs.swift.org/browse/TF-820). It adds a new top-level construct called "differentiability witnesses" to SIL and doesn't affect much existing code.

If there's no activity, I'll merge within a few days to unblock progress. Happy to address feedback at any time!

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
Looks good to me. If you're just adding data structures to support autodiff that have minimal interaction with the rest of the compiler like this, I would say don't wait for me.

**Comment by dan-zheng** [comment]
> Looks good to me. If you're just adding data structures to support autodiff that have minimal interaction with the rest of the compiler like this, I would say don't wait for me.

Got it, thank you!

If you don't mind, I'll continue to add you as a reviewer on SIL autodiff upstreaming patches, unless you have some other preference.

**Comment by dan-zheng** [comment]
The macOS `Python/python-lint` test failure seems [unrelated to this patch](https://github.com/apple/swift/pull/29636#issuecomment-581997070), so I'll merge now.

**Pull request merged**
